eth/filters/filter_system.go

```diff
@@ -14,172 +14,588 @@
-	LogFilter                          // new or removed log filter
-	PendingLogFilter                   // pending log filter
+	// UnknownFilter indicates an unkown filter type
+	UnknownFilter Type = iota
+	// BlockFilter queries for new blocks
+	BlockFilter
+	// PendingTxFilter queries pending transactions
+	PendingTxFilter
+	// LogFilter queries for new or removed (chain reorg) logs
+	LogFilter
+	// PendingLogFilter queries for logs for the pending block
+	PendingLogFilter
+
+	maxPendingHashes = 10240 // max buffer size of block/tx hashes before an filter is considered inactive and disabled
+	maxPendingLogs   = 10240 // max buffer size of logs before an filter is considered inactive and disabled
 )
 
-// FilterSystem manages filters that filter specific events such as
-// block, transaction and log events. The Filtering system can be used to listen
-// for specific LOG events fired by the EVM (Ethereum Virtual Machine).
-type FilterSystem struct {
-	filterMu sync.RWMutex
-	filterId int
+var (
+	errFilterNotFound           = errors.New("filter not found")
+	errCouldNotGenerateFilterID = errors.New("unable to generate filter id")
+	errInvalidFilterID          = errors.New("invalid filter id")
+	errUnableToUninstallFilter  = errors.New("unable to uninstall filter, retry later")
 
-	chainFilters      map[int]*Filter
-	pendingTxFilters  map[int]*Filter
-	logFilters        map[int]*Filter
-	pendingLogFilters map[int]*Filter
+	filterIDGenMu sync.Mutex
+	filterIDGen   = filterIDGenerator()
+)
 
-	// generic is an ugly hack for Get
-	generic map[int]*Filter
+type logsCallback func(FilterID, []Log)
 
-	sub event.Subscription
+// Type determines the kind of filter and is used to put the filter in to
+// the correct bucket when added.
+type Type byte
+
+// FilterID determines the type for a filter identifier.
+type FilterID [16]byte
+
+// MarshalJSON serializes a FilterID into its JSON representation.
+func (f FilterID) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fmt.Sprintf("0x%x", f))
 }
 
-// NewFilterSystem returns a newly allocated filter manager
-func NewFilterSystem(mux *event.TypeMux) *FilterSystem {
-	fs := &FilterSystem{
-		chainFilters:      make(map[int]*Filter),
-		pendingTxFilters:  make(map[int]*Filter),
-		logFilters:        make(map[int]*Filter),
-		pendingLogFilters: make(map[int]*Filter),
-		generic:           make(map[int]*Filter),
+// UnmarshalJSON parses a FilterID from its JSON representation.
+func (f *FilterID) UnmarshalJSON(data []byte) error {
+	// `"0x...."`
+	if len(data) != 36 || data[0] != '"' || data[35] != '"' || data[1] != '0' || (data[2] != 'x' && data[2] != 'X') {
+		return errInvalidFilterID
 	}
-	fs.sub = mux.Subscribe(
+
+	_, err := hex.Decode(f[:], data[3:35])
+	return err
+}
+
+// Log is a helper that can hold additional information about vm.Log
+// necessary for the RPC interface.
+type Log struct {
+	*vm.Log
+	Removed bool `json:"removed"`
+}
+
+type filter struct {
```

Comment:
The reason is that both do slightly different things, `Filter` is used to search for stored logs and `filter` is used by the filter system for state changes. We could combine them, but I prefer to do that in a separate PR.